### PR TITLE
Containerized build and flash

### PR DESCRIPTION
**Pull request by ErezBinyamin**
# What's new

- Added `docker_build` directory containing a Dockerfile and Makefile that simplify the build and flash process

# Verification 

1. Checkout/pull my fork
2. cd to `docker_build`
3. Physically connect a flipper device you wish to flash
4. Execute the command: `make flash`

# Checklist (For Reviewer)

- [ ] PR has description of feature/bug
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

**Comment by xMasterX** [comment]
I understand a purpose of this some time before when we had makefile and stuff like that, in that time we should have all dependencies installed by ourself, manually.

Now when we have fbt with fully automatic install and dependencies tracking, it also works on all platforms, mac, linux, windows

Why should this exist? Why do you bringing back deprecated way to build fw and for what reason?

You calling it will `simplify the build and flash process`, how exactly? 
Did you know that you can enter one command and it will do all without having whole docker running on your pc?
`./fbt COMPACT=1 DEBUG=0 FORCE=1 flash_usb_full` -> will build in release mode and flash it via usb into flipper, fully automatic, no extra steps, no need to have docker running, no need for extra software

Maybe I'm missing something and we actually need to bring old makefile back for some reason?

**Comment by ErezBinyamin** [comment]
Local builds vs docker builds are certainly a matter of preference. Here is a tradeoff explanation. Some developers may prefer to not install dependencies on their local machines (Other than docker). I do most of my development in containers and that helps me avoid package/dependency management issues.

What docker builds give you:
- Guarantee controlled build environment distributed to all developers
- No dependency conflicts on local machine

Drawbacks:
- The need to maintain a Dockerfile as new dependencies are introduced
- The pain of "having whole docker running on your pc"

Additionally:
The command `./fbt COMPACT=1 DEBUG=0 FORCE=1 flash_usb_full` is used in the Makefile I posted. You are correct that it makes the build process much easier. The value added with this dockerfile is simply the removal of the build from the local environment and moving it into a containerized and controlled one.

File: docker_build/Makefile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	touch .image_build
+	docker images
+
+check_port:
+	@echo "Checking if flipper device is connected somewhere in /dev ...."
+	[ -c $(PORT) ] && echo "Flipper Connected!"
+
+# Build and flash flipper
+flash: img check_port
+	docker run --device $(PORT):$(PORT) --env CUSTOM_FLIPPER_NAME="${CUSTOM_FLIPPER_NAME}" ${IMG} ./fbt COMPACT=1 DEBUG=0 FORCE=1 flash_usb_full
```
Comment (ErezBinyamin) [inline docker_build/Makefile:18]: Using `./fbt COMPACT=1 DEBUG=0 FORCE=1 flash_usb_full` do build and flash to flipper over USB

**Comment by xMasterX** [comment]
> Some developers may prefer to not install dependencies on their local machines (Other than docker)

You know how fbt works right now? it downloads a package into project folder, this package has all needed stuff

What dependencies you need to install to use fbt? Can you explain?
git and curl? thats a basic stuff, how you end up having docker but no git and curl on your machine?


> No dependency conflicts on local machine

What will conflict? can you explain this too?

**Comment by ErezBinyamin** [comment]
I did not know how fbt worked, I just am in the habit of containerized my builds. When I decided to build this firmware, I containerized it before even trying to build locally. 

You are correct. Fbt is a great solution. No need for the container.

Thank you for the explanation. Thanks for your patience. And my apologies for the waste of time.

-Erez

**Comment by ErezBinyamin** [comment]
Please close PR.

**Comment by xMasterX** [comment]
> I did not know how fbt worked, I just am in the habit of containerized my builds. When I decided to build this firmware, I containerized it before even trying to build locally.
> 
> You are correct. Fbt is a great solution. No need for the container.
> 
> Thank you for the explanation. Thanks for your patience. And my apologies for the waste of time.
> 
> -Erez

Sure no problems

As i said we had same exact solution with docker and makefile but in the past, when fbt was made this was removed since its all self containted now, all dependencies are located in project folder, so no extra software on pc except git and curl/wget

**Pull request closed without merging**